### DISCUSSION TITLE
SSO Setup

### DISCUSSION THREAD

**Discussion by CovantaDataManagement** [Budibase/budibase#3073] [Help] [answered]
I need some help to answer my infrastructure guys questions to complete SSO setup.

**We changed reply URL from https://localhost:10001/api/admin/auth/oidc/callback  to https://covforms.covanta.com/api/admin/auth/oidc/callback and it does not asking for admin consent but giving below error:

 

So please help us to get correct reply URL from Budibase support and also please help to configure below API permissions which need to requires.

 

[
![budi sso2](https://user-images.githubusercontent.com/62119887/137797708-3e9acb7b-8a55-431c-9916-abb5b9b95a4d.png)
![budi sso1 jpg](https://user-images.githubusercontent.com/62119887/137797710-decae01c-3915-49e6-a928-e16a76cfce45.png)
](url)**

**Comment by PClmnt** [comment]
Hey @CovantaDataManagement I'm just trying to debug through this, in Active Directory, have you defined the Platform as `Web` or `Single-page Application`? It should be Web.

**Reply by CovantaDataManagement** [reply]
We had the callback URL misconfigured between the two.  That is straightened out but now we have another issue after being redirected.

**Comment by Rory-Powell** [comment]
Hi @CovantaDataManagement , does the error occur before you are redirected to budibase?
The docs for configuring azure ad are located here https://docs.budibase.com/building-apps/admin-and-management/authentication/azure

**Reply by CovantaDataManagement** [reply]
We believe we have this sorted to the point that is redirects.  Now we are getting:
{"message":"Unauthorized","status":403}
From the URL:
_https://covforms.covanta.com_**/api/global/auth/oidc/callback**?code=0.ATgAC9mrSgl6F0qwTVLz9qj6sFi8JdKfQ_5NhaHyQ_wBRo44ABI.AQABAAIAAAD--DLA3VO7QrddgJg7Wevrt8rxcbI-B7_w5JsKi_fr6k1LPm6ZiEitft_2xJExsWiNrNXLlQuzvMvuipz4KSIl3fe3GeE-wZEo_3svwkXSFAao3KycyzmmpZzTtBj7phmASQSoEWCBKvPRGaV11WtX3kiyPlUFCpcAyCa9kmHn0JGgfJW2YHy0NwUWlYBN8Pjpsc-wg7YYfs5elPsN0bqPVnl_JdiUpHSNW6xi_soOfp11HLuloBffea4OL1dyqCjWiArtv0k3uRLXbU5fablRwEXoQ65s-d6ZglLBVic5ddHWTCnd0_afA3-19FWzgQl-bezgUjSG62AOr9Dg42pzEwqplVp7ZZ7PiZY3pAdlgIe9BynbVyz6MnKBXIapQ6fKqstTPoPFRPSmDPU2YTc5kUAySppVdMaDrgQbePUQp_RBk1AACGHZtL7QzcVM81EuqV-atAaLcgMjO7c0-A10Ymc9lpPAFw3SVYgHZzLcGk96WSTqlBXtsGHDwGAA62aizemrBDQmyTi7kUyMf1Cc7WsiMC58sdPDOFHTgWurGKkcsrcx4_944wxGrVJyF3ZF5V7G1tXXxXdveXyyFn0MrxKDjOQnGoxwar3h2czwc2ultZeE5mAV7ySwbuunvxmCWRDUn04LFiFkb7bXKayTJqoytY7VzCA6KFR0ltoMsthqKU-EJtQrIQrfPzRy8SUNTFxhLNkPuj0ttvTNTqe4IAA&state=%7b%22handle%22%3a%22juq27g6Lr3AexSONQ7nL5NnH%22%7d&session_state=3b909bbd-74dd-44d9-bfc3-201da8bce3bf#


Our AD admin says its successful on their end per the attached screenshot:
![budi sso4](https://user-images.githubusercontent.com/62119887/137966974-d62ccee5-1bd3-4387-b9bd-8a1215438017.png)

**Reply by mjashanks** [reply]
So, where are you seeing the error message? Is that displayed directly in the browser when you try to go to that URL?

bit of a stab in the dark here, but the URL ends with a `#`, which seems strange. Can you try to remove that to see what happens?

**Reply by shogunpurple** [reply]
@CovantaDataManagement it would be very helpful to see the logs from the server and worker when this error occurs. Which method have you used to self-host budibase? If you used the docker-compose setup, you should be able to get the logs by running:

`docker-compose logs worker-service`
and
`docker-compose logs app-service`

This will help provide more information, and allow us to debug. Thanks!

**Reply by CovantaDataManagement** [reply]
Thanks Martin.  Here are the logs I was able to pull.  The notable error I saw was Unauthorized - no public worker access.
Not sure what that means.
[apps-service.log](https://github.com/Budibase/budibase/files/7384254/apps-service.log)
[worker-service.log](https://github.com/Budibase/budibase/files/7384255/worker-service.log)

**Reply by CovantaDataManagement** [reply]
This is the actual URL:
https://covforms.covanta.com/builder/auth/login
When I actually choose SSO, the url with the # is passed and removing the # presents the same error.
https://covforms.covanta.com/api/global/auth/oidc/callback?code=0.ATgAC9mrSgl6F0qwTVLz9qj6sFi8JdKfQ_5NhaHyQ_wBRo44ABI.AQABAAIAAAD--DLA3VO7QrddgJg7WevrNyH4pk-GIeGvsae1fmXhmdwOYG1Mv09pBv82t5zmXARXc_CfBuVWsEDrNbMPSUdweK9sB50aD0gPJB9tyGKk_XVOgFBGdTnspramTMFXz1zZNwzZPIeS4DzO-jPvTersla3zQLb4NnNE1X_hNcPw-4PCPxkEL7N_7AVoZz9JUP84LCmjQl1lViWVYE36X0D7mfyclV_P3QzSiT7mrpxvPhm7v-uQHjIFYdsbFxIJcl502CjYzJWfz5kfIo1OtKQMwt9UAHwSHFUaxqcnw-7dPIkV4ww4gdSWbVTmGxzMrDSyIxFY_1ZGUSRP-cQsGIdsapfEg70pcnbzT6hw3EMvzAoJjlJvtgKrKLQtjoPNFn-4u1WmPtO2gNC-r9ztO4paEpE7Op8FGl5snUnO7N6QZfZu42zkUGU4og1vHLx4FXKeLKwxn7PIzmZL8kGvNkcTUq-w_K0MKSWrjQz_XQrVpR572DUpdJs_1lSyQYWceQcTKQASpil-0PiqhR-I3z8FDIJ48XaYvnzEI7Pg3nh4EIvMQmqMETu207sgM7xRqLPzRnsb4IPvdA33E_SgItoswi-thpftPOmcUlvN31DTyRwvfAggDIpi85cyFMXJfxAaEM1KbcI1ZbJJ9xh7FuueTpGWmC9w60GbTnubAGl9uG-75pwt_HMxFM3lpaiDKPjoBZ3cOKUXaB5T0f3kq1lxIAA&state=%7b%22handle%22%3a%226uP2YSv9OT%2bd4Z7dVXy%2fRVk7%22%7d&session_state=3b909bbd-74dd-44d9-bfc3-201da8bce3bf#

**Reply by shogunpurple** [reply]
Hi @CovantaDataManagement - thank you very much for the logs. That's super helpful. After reviewing, we noticed the following in the worker logs (the service responsible for auth)

```
bbworker  | Authentication error InternalOAuthError: failed to obtain access token
bbworker  |   oauthError: {
bbworker  |     statusCode: 401,
bbworker  |     data: '{"error":"invalid_client","error_description":"AADSTS7000215: Invalid client secret is provided.\\r\\nTrace ID: d5c95c40-5851-4974-9ba4-2b37894a4500\\r\\nCorrelation ID: ec672f5e-9a25-48cd-ba30-d9eda55e807d\\r\\nTimestamp: 2021-10-20 19:47:16Z","error_codes":[7000215],"timestamp":"2021-10-20 19:47:16Z","trace_id":"d5c95c40-5851-4974-9ba4-2b37894a4500","correlation_id":"ec672f5e-9a25-48cd-ba30-d9eda55e807d","error_uri":"https://login.microsoftonline.com/error?code=7000215"}'
bbworker  |   }
```

It looks like your client secret is not configured correctly. We noticed that there was a typo in the docs, that told users to put the client secret into the clientID field, which has been updated now, but we don't believe is the problem here.

What is more likely is that the client secret has not been copied correctly from Azure AD - the UX is a little confusing. Ensure that the **Value** of the secret is what is copied, rather than the Secret ID.
![Frame 1(1)](https://user-images.githubusercontent.com/11256663/138247509-e29728d4-fb6b-4259-8d26-256e13f4c2d2.png)

Could you please confirm that the client secret is set up as expected, and that it matches the one in budibase? If so, we can continue to debug from there.

**Comment by Rory-Powell** [comment, marked as answer]
We resolved this on a call with @CovantaDataManagement based on the answer from @shogunpurple here https://github.com/Budibase/budibase/discussions/3073#discussioncomment-1513200
